File: typescript-automation/src/driver/ports/browserPort.ts
export interface BrowserPort {
  // Navigation
  navigateTo(url: string): Promise<void>;
  navigateBack(): Promise<void>;
  currentUrl(): Promise<string>;

  // Element state
  isVisible(css: string): Promise<boolean>;
  isPresent(css: string): Promise<boolean>;
  isEnabled(css: string): Promise<boolean>;
  isSelected(css: string): Promise<boolean>;
  count(css: string): Promise<number>;

  // Element content
  text(css: string): Promise<string>;
  attribute(css: string, attr: string): Promise<string>;

  // Nth-element queries (flat list, zero-based index)
  isNthEnabled(css: string, index: number): Promise<boolean>;
  isNthSelected(css: string, index: number): Promise<boolean>;
  nthAttribute(css: string, index: number, attr: string): Promise<string>;
  nthText(css: string, index: number): Promise<string>;

  // Scoped queries — within the nth element matching parentCss
  isSelectedWithin(parentCss: string, parentIndex: number, childCss: string): Promise<boolean>;
  attributeWithin(parentCss: string, parentIndex: number, childCss: string, attr: string): Promise<string>;

  // Actions
  click(css: string): Promise<void>;
  clickNth(css: string, index: number): Promise<void>;
  clickXpath(xpath: string): Promise<void>;
  sendKeys(css: string, text: string, submitAfter: boolean): Promise<void>;
  setReactInputValue(css: string, value: string): Promise<void>;

  // JavaScript extraction
  extractAllViaScript(script: string): Promise<ReadonlyArray<Readonly<Record<string, string>>>>;
  executeScript(script: string, ...args: unknown[]): Promise<unknown>;

  // Waits
  waitUntilVisible(css: string): Promise<void>;
  waitUntilPresent(css: string): Promise<void>;
  waitUntilCountMoreThan(css: string, count: number): Promise<void>;
  waitUntilUrlContains(fragment: string): Promise<void>;
  waitUntilUrlMatches(regex: string): Promise<void>;
  waitUntilAttributeChanges(css: string, index: number, attr: string, previousValue: string): Promise<void>;
  waitUntilAnyPresent(...cssList: string[]): Promise<void>;
  waitUntilCondition(condition: () => Promise<boolean>, timeoutMs: number): Promise<void>;
  tryWaitUntilPresent(css: string, timeoutMs: number): Promise<boolean>;
}

File: python-automation/src/automation/driver/ports/browser_port.py
from abc import ABC, abstractmethod
from collections.abc import Callable
from typing import Any


class BrowserPort(ABC):

    @abstractmethod
    def navigate_to(self, url: str) -> None: ...

    @abstractmethod
    def navigate_back(self) -> None: ...

    @abstractmethod
    def current_url(self) -> str: ...

    @abstractmethod
    def is_visible(self, css: str) -> bool: ...

    @abstractmethod
    def is_present(self, css: str) -> bool: ...

    @abstractmethod
    def is_enabled(self, css: str) -> bool: ...

    @abstractmethod
    def is_selected(self, css: str) -> bool: ...

    @abstractmethod
    def count(self, css: str) -> int: ...

    @abstractmethod
    def text(self, css: str) -> str: ...

    @abstractmethod
    def attribute(self, css: str, attr: str) -> str: ...

    @abstractmethod
    def is_nth_enabled(self, css: str, index: int) -> bool: ...

    @abstractmethod
    def is_nth_selected(self, css: str, index: int) -> bool: ...

    @abstractmethod
    def nth_attribute(self, css: str, index: int, attr: str) -> str: ...

    @abstractmethod
    def nth_text(self, css: str, index: int) -> str: ...

    @abstractmethod
    def is_selected_within(self, parent_css: str, parent_index: int, child_css: str) -> bool: ...

    @abstractmethod
    def attribute_within(
        self, parent_css: str, parent_index: int, child_css: str, attr: str
    ) -> str: ...

    @abstractmethod
    def click(self, css: str) -> None: ...

    @abstractmethod
    def click_nth(self, css: str, index: int) -> None: ...

    @abstractmethod
    def click_xpath(self, xpath: str) -> None: ...

    @abstractmethod
    def send_keys(self, css: str, text: str, submit_after: bool) -> None: ...

    @abstractmethod
    def set_react_input_value(self, css: str, value: str) -> None: ...

    @abstractmethod
    def extract_all_via_script(self, script: str) -> list[dict[str, str]]: ...

    @abstractmethod
    def execute_script(self, script: str, *args: Any) -> Any: ...

    @abstractmethod
    def wait_until_visible(self, css: str) -> None: ...

    @abstractmethod
    def wait_until_present(self, css: str) -> None: ...

    @abstractmethod
    def wait_until_count_more_than(self, css: str, count: int) -> None: ...

    @abstractmethod
    def wait_until_url_contains(self, fragment: str) -> None: ...

    @abstractmethod
    def wait_until_url_matches(self, regex: str) -> None: ...

    @abstractmethod
    def wait_until_attribute_changes(
        self, css: str, index: int, attr: str, previous_value: str
    ) -> None: ...

    @abstractmethod
    def wait_until_any_present(self, *css_list: str) -> None: ...

    @abstractmethod
    def wait_until_condition(
        self, condition: Callable[[], bool], timeout_seconds: int
    ) -> None: ...

    @abstractmethod
    def try_wait_until_present(self, css: str, timeout_seconds: int) -> bool: ...

    @abstractmethod
    def close(self) -> None: ...
